clients/drdisasm/drdisasm.cpp

```diff
@@ -0,0 +1,158 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#define DR_FAST_IR 1
+#include "dr_api.h"
+#include "droption.h"
+#include <iostream>
+#include <sstream>
+
+namespace {
+
+// XXX i#1684: We want cross-arch decoding support so a single build can decode
+// AArchXX and x86.  For now, a separate build is needed.
+#ifdef X86_64
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "x64",
+                                "Decodes using the specified mode: 'x64' or 'x86'.",
+                                "Decodes using the specified mode: 'x64' or 'x86'.");
+#elif defined(ARM)
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "arm",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.");
+#endif
+droption_t<bool> op_show_bytes(DROPTION_SCOPE_FRONTEND, "show_bytes", true,
+                               "Display the instruction encoding bytes.",
+                               "Display the instruction encoding bytes.");
+
+static bool
+parse_bytes(std::string token, std::vector<byte> &bytes)
+{
+    // Assume everything is hex.
+    // Assume consecutive values should be split and endianized.
+    uint64 entry;
+    std::stringstream stream;
+    stream << std::hex << token;
+    if (!(stream >> entry))
+        return false;
+    do {
+        bytes.push_back(entry & 0xff);
+        entry >>= 8;
+    } while (entry > 0);
+    return true;
+}
+};
+
+int
+main(int argc, const char *argv[])
+{
+    int last_index;
+    std::string parse_err;
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, argv, &parse_err,
+                                       &last_index)) {
+        std::cerr << "Usage error: " << parse_err << "\nUsage:\n " << argv[0]
+                  << " [options] <hexadecimal bytes to decode as args or stdin>\n"
```

Comment:
Mention that the byte sequence needs to be tokenizable in the usage error message. For instance ./drdisasm 0x140x320x110x99 causes issues (however: ./drdisasm 14321199 and 0x14321199 are fine).

Reply:
I did not consider the hex literal style you are showing there: normally that would be a backslash-x.  Have you seen it the way you're showing with no backslash in other places?  Is that something to support?  I figured you'd just use spaces to separate.  See also above where 0x14321199 is interpreted as "99 11 32 14" (i.e., little-endian word), and so is perhaps not the same as what would be meant by \x14\x32\x11\x99 which I would assume should be "14 32 11 99".

Reply:
Yes I agree. When reading the code I was still in the mind-set of using little-endian... Just that, since I always use llvm-mc for disassembly, which separates hex bytes in the form of 0xXX via spaces, I assumed the the user shouldn't remove the space separator, hence being tokenisable.

Reply:
Wow I misread your 0x140x320x110x99: ok so that was 0x14 0x32 0x11 0x99 with the spaces removed.  I thought it was trying to be the C string hex escape before each byte.  Anyway we are all on the same page now.

Reply:
I added this to the usage message:
```
                  << "Bytes do not need leading 0x.  Single-token multi-byte values are "
                  << "assumed to be machine-endian words.\n"
```

Reply:
Yea, I think that is clearer. Thanks for the update.